pageserver/src/tenant.rs

```diff
@@ -2874,24 +2917,34 @@ impl Tenant {
                 }
             }
 
-            let cutoff = timeline
-                .get_last_record_lsn()
-                .checked_sub(horizon)
-                .unwrap_or(Lsn(0));
-
             let branchpoints: Vec<Lsn> = all_branchpoints
                 .range((
                     Included((timeline_id, Lsn(0))),
                     Included((timeline_id, Lsn(u64::MAX))),
                 ))
                 .map(|&x| x.1)
                 .collect();
-            let cutoffs = timeline.find_gc_cutoffs(cutoff, pitr, cancel, ctx).await?;
 
-            *timeline.gc_info.write().unwrap() = GcInfo {
-                retain_lsns: branchpoints,
-                cutoffs,
-            };
+            {
+                let mut target = timeline.gc_info.write().unwrap();
+
+                match gc_cutoffs.remove(&timeline_id) {
+                    Some(cutoffs) => {
+                        *target = GcInfo {
+                            retain_lsns: branchpoints,
+                            cutoffs,
+                        };
+                    }
+                    None => {
+                        // reasons for this being unavailable:
+                        // - this timeline was created while we were finding cutoffs
+                        // - lsn for timestamp search fails for this timeline repeatedly
+                        //
+                        // in both cases, refreshing the branchpoints is correct.
+                        target.retain_lsns = branchpoints;
```

Comment:
Is this the reason why it's correct? Also s/correct/safe/ ?

```suggestion
                        // in both cases, updating the branchpoints while leaving
                        // previous iteration's cutoffs is safe because if we don't
                        // change the cutoffs, we don't risk removing any data we shouldn't.
                        //
                        // It's not a situation we want to entertain in the long run, because
                        // not GCing means _someone_ has to pay for that storage.
                        // But, for now, we'll be optimistic, that the next iteration will
                        // reach the `Some()` case above.
                        target.retain_lsns = branchpoints;
```

Reply:
I think it's more correct to always update the branch points than just 
safe, but in general, we should not get into situations where latest_gc_cutoff has moved past the branch points if we are constantly failing the pitr_cutoff finding.

Branch points are used for synthetic size calculation in addition to GC. Synthetic size will grow from the inability to remove layer files, as the pitr_cutoff will stay fixed.